File: bin/load_reverts.php
<?php
echo "<pre>";

//Coleta lista de edições
$edits_statement = "
    SELECT
        `diff`
    FROM
        `{$contest['name_id']}__edits`
    WHERE
        `valid_user` IS NOT NULL AND
        `reverted` IS NULL
";
$edits_query = mysqli_prepare($con, $edits_statement);
mysqli_stmt_execute($edits_query);
$edits_result = mysqli_stmt_get_result($edits_query);

//Verifica se existem edições cadastradas no banco de dados
$rows = mysqli_num_rows($edits_result);
if ($rows == 0) { die("No edits"); }

//Prepara query para atualizações no banco de dados
$update_statement = "
    UPDATE
        `{$contest['name_id']}__edits`
    SET
        `reverted` = '1'
    WHERE
        `diff` = ?
";
$update_query = mysqli_prepare($con, $update_statement);
mysqli_stmt_bind_param($update_query, "i", $diff);

//Loop para análise de cada edição
while ($row = mysqli_fetch_assoc($edits_result)) {

    //Coleta tags da revisão
    $revisions_api_params = [
        "action"    => "query",
        "format"    => "php",
        "prop"      => "revisions",
        "rvprop"    => "sha1|tags",
        "revids"    => $row["diff"]
    ];
    $revisions_api = file_get_contents($contest['api_endpoint']."?".http_build_query($revisions_api_params));
    $revisions_api = unserialize($revisions_api)['query'];
    $revision = end($revisions_api['pages'])['revisions']['0'];

    //Marca edição caso tenha sido revertida ou eliminada
    if (
        isset($revisions_api['badrevids'])
        || isset($revision['sha1hidden'])
        || in_array('mw-reverted', $revision['tags'])
    ) {
        $diff = $row["diff"];
        mysqli_stmt_execute($update_query);
        if (mysqli_stmt_affected_rows($update_query) != 0) {
            echo "Marcada edição {$row["diff"]} como revertida.<br>";
        }
    }
}

//Encerra conexão
mysqli_stmt_close($update_query);
mysqli_close($con);
echo "Concluido! (3/3)";

File: contests/management/commands/load_reverts.py
import requests
import time
from datetime import datetime
from django.core.management.base import BaseCommand
from django.db import connection
from contests.models import Contest, Article, Edit, Qualification
from django.db.models import OuterRef, Subquery

class Command(BaseCommand):
    help = 'Update reverts'

    def add_arguments(self, parser):
        parser.add_argument('contest', type=str, help="Nome ID do concurso")

    def handle(self, *args, **options):
        contest_name_id = options.get('contest')
        contest = Contest.objects.get(name_id=contest_name_id)

        # Coleta lista de edições ativas no concurso
        diffs = Edit.objects.filter(contest=contest, last_qualification__status=1).values_list('diff', flat=True)

        # Loop para análise de cada edição
        for diff in diffs:

            # Coleta tags da revisão
            revisions_api_params = {
                "action": "query",
                "format": "json",
                "prop": "revisions",
                "rvprop": "sha1|tags",
                "revids": diff,
            }
            revisions_api = requests.get(contest.api_endpoint, params=revisions_api_params).json()
            revisions_api_query = revisions_api.get('query', {})
            revision = None

            if 'pages' in revisions_api_query:
                self.stdout.write(f"Revisão: {diff}")
                revision_page = next(iter(revisions_api_query['pages'].values()), None)
                revision = revision_page['revisions'][0] if revision_page and 'revisions' in revision_page else None

            # Marca edição caso tenha sido revertida ou eliminada
            if (
                'badrevids' in revisions_api_query
                or (revision and 'sha1hidden' in revision)
                or (revision and 'mw-reverted' in revision['tags'])
            ):
                new_qualification = Qualification.objects.create(
                    contest=contest,
                    diff=Edit.objects.get(contest=contest, diff=diff),
                    status=0,
                )
                Edit.objects.filter(contest=contest, diff=diff).update(last_qualification=new_qualification)
                self.stdout.write(f"Marcada edição {diff} como revertida.")

        # Encerra script
        self.stdout.write("Concluído! (3/3)")
